File: impl/ts/src/runtime/fetch.ts
/**
 * SEMP_FETCH handler. Mirrors impl/go/internal/runtime/fetch.go.
 *
 * SEMP_FETCH is a demo-only client/home-server inbox pull. The
 * dispatcher routes it through the `onUnknown` hook because the
 * semp-ts session.runDispatcher does not have a dedicated
 * onFetch handler.
 *
 * @module
 */

import {
  newFetchResponse,
  FetchType,
  type Inbox as MemInbox,
  type FetchRequest,
} from "@sempdev/semp/delivery";
import type { Logger } from "pino";

/** Drain the inbox for `identity` and return the wire response. */
export function handleFetch(
  raw: Uint8Array,
  inbox: MemInbox,
  identity: string,
  logger: Logger | null,
): Uint8Array {
  const req = JSON.parse(new TextDecoder().decode(raw)) as FetchRequest;
  if (req.type !== FetchType || req.step !== "request") {
    throw new Error(
      `unexpected fetch type/step: ${req.type}/${req.step}`,
    );
  }
  const queued = inbox.drain(identity);
  const out = queued.map((b) => Buffer.from(b).toString("base64"));
  logger?.info(
    { identity, count: out.length },
    "fetch returned envelopes",
  );
  return new TextEncoder().encode(JSON.stringify(newFetchResponse(out)));
}

File: impl/go/internal/runtime/fetch.go
package runtime

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log/slog"

	"semp.dev/semp-go/delivery"
	"semp.dev/semp-go/transport"
)

// handleFetch returns every waiting envelope for the authenticated
// identity. SEMP_FETCH is a demo-only extension (HANDSHAKE.md section
// 4.6 leaves the wakeup mechanism out of scope).
func handleFetch(ctx context.Context, conn transport.Conn, raw []byte, inbox *delivery.Inbox, identity string, logger *slog.Logger) error {
	var req delivery.FetchRequest
	if err := json.Unmarshal(raw, &req); err != nil {
		return fmt.Errorf("parse fetch request: %w", err)
	}
	if req.Type != delivery.FetchType || req.Step != delivery.FetchStepRequest {
		return fmt.Errorf("unexpected fetch type/step: %s/%s", req.Type, req.Step)
	}
	queued := inbox.Drain(identity)
	out := make([]string, 0, len(queued))
	for _, payload := range queued {
		out = append(out, base64.StdEncoding.EncodeToString(payload))
	}
	if logger != nil {
		logger.Info("fetch returned envelopes",
			"identity", identity,
			"count", len(out),
		)
	}
	return sendJSON(ctx, conn, delivery.NewFetchResponse(out))
}
